### PR TITLE
James 1901 - Report metrics in ElasticSearch

### PR DESCRIPTION
**Pull request by chibenwa**
No tests yet. It's thought to not chrash if config file is absent.

Also as it uses http, the dependency added is super light. Also as it uses HTTP, james still start without ES (reports just fails, error log).

The report are done in james-metrics-2016-12

screenshot from 2016-12-30 18-57-02

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        Multibinder.newSetBinder(binder(), ConfigurationPerformer.class).addBinding().to(ESMetricReporterStarter.class);
+    }
+
+    @Provides
+    public ESReporterConfiguration provideConfiguration(FileSystem fileSystem) throws ConfigurationException {
+        try {
+            PropertiesConfiguration propertiesReader = getPropertiesConfiguration(fileSystem);
+
+            if (propertiesReader.getBoolean("elasticsearch.metrics.reports.enabled", DEFAULT_DISABLE)) {
+                return ESReporterConfiguration.enabled(
```
Comment (strokyl) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java]: I don't know Guice (neither Jame enough to be sure my suggestion make sens). But it look like you can load property files in Guice so you can easily inject them latter : http://stackoverflow.com/questions/3071891/guice-and-properties-files.

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+
+public class ESReporterConfiguration {
+
+    public static final boolean ENABLED = true;
+    public static final boolean DISABLED = !ENABLED;
+    public static final String DEFAULT_INDEX = "james-metrics";
+    public static final long DEFAULT_period_in_second = 60L;
```
Comment (aduprat) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java]: Upper case

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import com.google.common.base.Throwables;
+
+public class ESMetricReporter {
+
+    private final Optional<ElasticsearchReporter> reporter;
+    private final ESReporterConfiguration esReporterConfiguration;
+
+    public ESMetricReporter(ESReporterConfiguration esReporterConfiguration, MetricRegistry registry) {
+        if (esReporterConfiguration.isEnabled()) {
+            try {
```
Comment (aduprat) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: Can you extract this.

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    protected void configure() {
+        Multibinder.newSetBinder(binder(), ConfigurationPerformer.class).addBinding().to(ESMetricReporterStarter.class);
+    }
+
+    @Provides
+    public ESReporterConfiguration provideConfiguration(FileSystem fileSystem) throws ConfigurationException {
+        try {
+            PropertiesConfiguration propertiesReader = getPropertiesConfiguration(fileSystem);
+
+            if (propertiesReader.getBoolean("elasticsearch.metrics.reports.enabled", DEFAULT_DISABLE)) {
```
Comment (aduprat) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java]: Extract please: isMetricEnable

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java [hunk 5]
```diff
@@ -66,8 +69,8 @@ protected void configure() {
     protected Client provideClientProvider(FileSystem fileSystem, AsyncRetryExecutor executor) throws ConfigurationException, FileNotFoundException, ExecutionException, InterruptedException {
         PropertiesConfiguration propertiesReader = new PropertiesConfiguration(fileSystem.getFile(FileSystem.FILE_PROTOCOL_AND_CONF + "elasticsearch.properties"));
```
Comment (aduprat) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java]: ES_CONFIG_FILE

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java (hunk 1 above)
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/server/ESMetricReporterModule.java]: Yes, but we usually try to limit the use of libraries to what we want and for the time being, we want to write our configuration reading.
I'll probably change once configuration reading is less of a mess.

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java (hunk 2 above)
Comment (Laura-Royet) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESReporterConfiguration.java]: Done

File: server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java (hunk 3 above)
Comment (Laura-Royet) [inline server/container/metrics/metrics-dropwizard/src/main/java/org/apache/james/metrics/dropwizard/ESMetricReporter.java]: Done

**Comment by Laura-Royet** [comment]
The comments I took into account are here: https://github.com/linagora/james-project/pull/547

**Pull request closed without merging**